File: packages/feedback/src/count/civ-count.ts
import { html, nothing } from 'lit';
import { customElement, property } from 'lit/decorators.js';
import { ifDefined } from 'lit/directives/if-defined.js';
import { CivBaseElement } from '@civui/core';

export type CountIntent = 'info' | 'warning' | 'error' | 'success' | 'neutral';
export type CountEmphasis = 'primary' | 'secondary' | 'tertiary';
export type CountLive = 'off' | 'polite' | 'assertive';

/**
 * CivUI Count
 *
 * A numeric annotation — match counts on filter chips, item counts in
 * navigation, notification counters, etc. Lighter chrome than
 * `civ-badge` so it sits inside other components without competing
 * with their visual weight.
 *
 * **Styles** (`count-style`):
 * - `secondary` (default) — tinted background pill with semantic color text.
 *   Use inside chips, list items, nav links.
 * - `primary` — filled rounded pill with white text. Use for standalone
 *   or notification-style counters.
 * - `tertiary` — bare text with parentheses, e.g. "(12)". Minimal
 *   emphasis for inline counts.
 *
 * **Position** is independent of style: set `overlay` to absolutely
 * position the count in the top-end corner of a relatively-positioned
 * parent (use the `civ-badge-anchor` utility class), regardless of
 * whether the style is `secondary` or `primary`.
 *
 * **Accessibility:** counts are static annotations by default
 * (`live="off"`). Set `live="polite"` for inbox-style counters that
 * change as data updates and should announce.
 *
 * @element civ-count
 *
 * @prop {number | null} count - Numeric value. Render is skipped when null, undefined, or NaN.
 * @prop {number} max - Overflow threshold; values above render as `{max}+` (default 99)
 * @prop {CountIntent} intent - Semantic color (default 'neutral')
 * @prop {CountEmphasis} emphasis - Emphasis: 'secondary' (text only, default) or 'primary' (filled pill)
 * @prop {string} spacing - Padding size: 'default' or 'sm'
 * @prop {boolean} overlay - Position absolutely in the top-end corner of a relative parent
 * @prop {CountLive} live - aria-live politeness: 'off' (default), 'polite', or 'assertive'
 *
 * @example
 * ```html
 * <civ-count count="24"></civ-count>
 * <civ-count count="3" intent="error" emphasis="primary"></civ-count>
 * <civ-count count="150" max="99" intent="info" overlay live="polite"></civ-count>
 * ```
 */
@customElement('civ-count')
export class CivCount extends CivBaseElement {
  /** Numeric value to render. Render is skipped when null, undefined, or NaN. */
  @property({ type: Number }) count: number | null = null;

  /** Overflow threshold; counts above render as "{max}+". */
  @property({ type: Number }) max = 99;

  /** Semantic color variant. */
  @property({ type: String }) intent: CountIntent = 'neutral';

  /** Emphasis: 'primary' (filled pill) or 'secondary' (text only, default). */
  @property({ type: String, attribute: 'emphasis' }) emphasis: CountEmphasis = 'secondary';

  /** Padding size: 'default' or 'sm' for compact layouts. */
  @property({ type: String }) spacing: 'default' | 'sm' = 'default';

  /** Notification-overlay positioning (top-end of a relative parent). */
  @property({ type: Boolean, reflect: true }) overlay = false;

  /** aria-live politeness for dynamic counters. Default `off` (static). */
  @property({ type: String }) live: CountLive = 'off';

  private get _displayCount(): string {
    return this.count! > this.max ? `${this.max}+` : String(this.count);
  }

  override render() {
    if (this.count === null || this.count === undefined || Number.isNaN(this.count)) {
      return nothing;
    }

    const classes = [
      'civ-count',
      `civ-count--${this.intent}`,
      `civ-count--style-${this.emphasis}`,
    ];
    if (this.spacing === 'sm') classes.push('civ-count--sm');
    if (this.overlay) classes.push('civ-count--overlay');

    const ariaLive = this.live === 'off' ? undefined : this.live;
    const role = this.live === 'off' ? undefined : 'status';

    return html`
      <span
        class="${classes.join(' ')}"
        role="${ifDefined(role)}"
        aria-live="${ifDefined(ariaLive)}"
      >${this.emphasis === 'tertiary' ? `(${this._displayCount})` : this._displayCount}</span>
    `;
  }
}

declare global {
  interface HTMLElementTagNameMap {
    'civ-count': CivCount;
  }
}

File: packages/ios/Sources/CivUI/CivCount.swift
// CivUI — CivCount for SwiftUI
// Numeric annotation following government design system patterns.

import SwiftUI

/// Accessible numeric counter for government applications.
///
/// Renders a numeric annotation — match counts on filter chips, item counts
/// in navigation, notification counters. Lighter chrome than `civ-badge`.
/// Mirrors the web `civ-count` component.
///
/// Usage:
/// ```swift
/// CivCount(count: 24)
/// CivCount(count: 3, variant: "error", emphasis: "primary")
/// ```
public struct CivCount: View {
    // MARK: - Properties

    /// Numeric value. Render is skipped when nil.
    public var count: Int?

    /// Overflow threshold; counts above render as "max+".
    public var max: Int

    /// Semantic color variant (e.g., "neutral", "error", "success").
    public var intent: String

    /// Emphasis style: "primary" (filled pill) or "secondary" (text only).
    public var emphasis: String

    /// Spacing variant.
    public var spacing: String

    /// Whether the count is positioned as a notification overlay.
    public var overlay: Bool

    /// aria-live politeness: "off", "polite", or "assertive".
    public var live: String

    // MARK: - Internal State

    @Environment(\.colorScheme) private var colorScheme

    // MARK: - Initializer

    public init(
        count: Int? = nil,
        max: Int = 99,
        intent: String = "neutral",
        emphasis: String = "secondary",
        spacing: String = "default",
        overlay: Bool = false,
        live: String = "off"
    ) {
        self.count = count
        self.max = max
        self.intent = intent
        self.emphasis = emphasis
        self.spacing = spacing
        self.overlay = overlay
        self.live = live
    }

    // MARK: - Body

    public var body: some View {
        EmptyView()
    }

    // MARK: - Color Helper

    private func adaptiveColor(light: Color, dark: Color) -> Color {
        colorScheme == .dark ? dark : light
    }
}

// MARK: - Preview

#if DEBUG
struct CivCount_Previews: PreviewProvider {
    static var previews: some View {
        VStack(spacing: 8) {
            CivCount(count: 24)
            CivCount(count: 3, variant: "error", emphasis: "primary")
            CivCount(count: 150, variant: "info", overlay: true, live: "polite")
        }.padding()
    }
}
#endif
